psi4/src/core.cc

```diff
@@ -781,33 +792,19 @@ bool py_psi_set_local_option_array(std::string const& module, std::string const&
         if (data.type() == "array")
             Process::environment.options.set_array(module, nonconst_key);
     }
-    size_t size = len(values);
-    for (int n = 0; n < size; ++n) {
-        try {
-            py::list l = values[n].cast<py::list>();
-            DataType *newentry = Process::environment.options.set_local_array_array(module, nonconst_key, entry);
-            // Now we need to recurse, to fill in the data
-            py_psi_set_local_option_array(module, key, l, newentry);
-        }
-        catch (py::cast_error e) {
-            // This is not a list; try to cast to a string
-            try {
-                std::string s = values[n].cast<std::string>();
-                Process::environment.options.set_local_array_string(module, nonconst_key, s, entry);
-            } catch(py::cast_error e) {
-                try {
-                    // This is not a list or string; try to cast to an integer
-                    int i = values[n].cast<int>();
-                    Process::environment.options.set_local_array_int(module, nonconst_key, i, entry);
-                } catch(py::cast_error e) {
-                    // This had better be castable to a float.  We don't catch the exception here
-                    // because if we encounter one, something bad has happened
-                    double f = values[n].cast<double>();
-                    Process::environment.options.set_local_array_double(module, nonconst_key, f, entry);
-                }
-            }
-
+    Options& options=Process::environment.options;
+    for(py::handle oi: values){
+        if(can_cast<py::list>(oi)){
+            DataType* newentry = options.set_local_array_array(module,nonconst_key, entry);
+            py_psi_set_local_option_array(module,key,oi.cast<py::list>(),newentry);
         }
+        else if(can_cast<std::string>(oi))
+            options.set_local_array_string(module, nonconst_key, oi.cast<std::string>(), entry);
+        else if(can_cast<int>(oi))
+            options.set_local_array_int(module,nonconst_key, oi.cast<int>(), entry);
+        else if(can_cast<double>(oi))
+            options.set_local_array_double(module,nonconst_key, oi.cast<double>(), entry);
+        else throw PSIEXCEPTION("unrecognized type in py_psi_set_local_option_array");
```

Comment:
All this looks tidier, to be sure, but what motivated the change? Any more resistant to pb11 updates?